app/src/main/java/com/example/campusguide/directions/DirectionsFlow.kt

```diff
@@ -0,0 +1,46 @@
+package com.example.campusguide.directions
+
+import android.content.Intent
+import androidx.appcompat.app.AppCompatActivity
+import com.example.campusguide.DirectionsActivity
+import com.example.campusguide.location.FusedLocationProvider
+import com.example.campusguide.utils.permissions.Permissions
+import com.google.maps.model.LatLng
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+class DirectionsFlow constructor(private val activity: AppCompatActivity, private val permissions: Permissions, private val locationProvider: FusedLocationProvider) {
+    fun startFlow(origin: String?, destination: String?) {
+        GlobalScope.launch {
+            val finalDestination = destination ?: getDestination()
+            val finalOrigin = origin ?: getOrigin()
+
+            val intent = Intent(activity, DirectionsActivity::class.java).apply {
+                putExtra("Origin", finalOrigin)
+                putExtra("Destination", finalDestination)
+            }
+            activity.startActivity(intent)
+        }
+    }
+
+    private suspend fun getDestination() = suspendCoroutine<String> { cont ->
+        val chooseDestinationOptions = ChooseDestinationOptions { destination ->
+            val destinationLatLng = LatLng(destination.latitude, destination.longitude)
+            cont.resume(destinationLatLng.toString())
+        }
+        chooseDestinationOptions.show(
+            activity.supportFragmentManager,
+            "chooseDestinationOptions"
+        )
+    }
+
+    private suspend fun getOrigin() = suspendCoroutine<String> { cont ->
```

Comment:
what do these 2 functions do?

Reply:
getOrigin and getDestination

Reply:
Each creates their popup that asks the user to enter origin / destination (current location, calendar event, search for location, etc). Thanks to the power of coroutines, the function will return once the popup is closed :)